text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+
+Because we're hyper-focused on backward- and forward-compatibility, there is no harm in progressively converting some addons to v2 (which provides immediate benefits in terms of build performance and reduced fragility under Embroider) while others need to stay as v1 until we offer the features they need.
+
+## Package Public API Overview
+
+The format we are about to describe _is a publication format_. Not necessarily an authoring format. By separating the two, we make it easier to evolve the authoring format without breaking ecosystem-wide compatibility. The publication format is deliberately more explicit and less dynamic that what we may want for an authoring format.
+
+First, here’s the list of things a v2 package can provide. More detail on each of these will follow:
+
+- **Own Javascript**: javascript and templates under the package’s own namespace (the v1 equivalent is `/addon/**/*.{js,hbs}/`)
+- **App Javascript**: javascript and templates that must be merged with the consuming app’s namespace (the v1 equivalent is `/app/**/*.{js,hbs}`). Other RFCs are working to move Ember away from needing this feature, but we are not gated on any of those and fully support **App Javascript**.
+- **Fastboot Javascript**: just like **App Javascript**, but only applies when running in Fastboot.
+- **CSS**: available for `@import` by other CSS files (both in the same package and across packages) and by ECMA `import` directives in Javascript modules (both in the same package and across packages).
+- **Implicit Dependencies**: scripts, modules, and stylesheets that should be implicitly included in the app or the app's tests whenever this addon is active. This is a backward-compatibility feature.
+- **Renaming Rules**: allow a package to declare that some of its modules should be available at different import paths than their real, resolvable path. This is a backward-compatibility feature that new addons should not use.
+- **Externals**: allows a package to declare that it imports some things that are not **allowed dependencies**. Instead of being resolved at build time, **externals** are deferred until runtime and get handled by the traditional `loader.js` `require()`.
+- **Assets**: any files that must be available in the final built application directory such that they have public URLs (typical examples are images and fonts).
+- **Build Hooks**: code that runs within Node at application build time. The v1 equivalent is an addon's `index.js` file.
+
+## Own Javascript
+
+The public `main` (as defined in `package.json`) of a v2 package points to its **Own Javascript**. The code is formatted as ES modules using ES latest features, meaning: stage 4 features only, unless otherwise explicitly mentioned elsewhere in this spec. (Addon authors can still use whatever custom syntax they want, but those babel plugins must run before publication to NPM.)
```

Comment:
Given that decorators are on the one hand still not stable (i.e. stage 2), but so pivotal already in the Ember Octane (and before) programming model, shouldn't we treat them as an exception to the rule here, and guarantee v2 addons that they will be transpiled, so they don't have to bother setting up a prepublish build step just because of the use of decorators? (which I believe is the only non-standard language feature broadly in use)

Reply:
Correct. I just pushed an update that tries to make this clearer by breaking out a separate section of the document for "Ember Standard Language", which explicitly allows decorators and some other features.